src/noit_metric_director.c

```diff
@@ -280,47 +413,70 @@ noit_metric_director_flush(eventer_t e) {
 
 static void
 distribute_metric(noit_metric_message_t *message) {
-  void *vhash, *vinterests;
-  caql_cnt_t *interests = NULL;
-  char uuid_str[UUID_STR_LEN + 1];
-  mtev_uuid_unparse_lower(message->id.id, uuid_str);
-  if(mtev_hash_retrieve(&id_level, (const char *) &message->id.id, UUID_SIZE,
-      &vhash)) {
-    if(mtev_hash_retrieve((mtev_hash_table *) vhash, message->id.name,
-        message->id.name_len_with_tags, &vinterests)) {
-      interests = vinterests;
-      distribute_message_with_interests(interests, message);
+  int i;
+  static uuid_t uuid_zero = { 0 };
+  void *vhash, **vinterests;
+  interest_cnt_t has_interests = 0;
+  interest_cnt_t interests[nthreads];
+  memset(interests, 0, sizeof(interests));
+
+
+  mtev_memory_begin();
+  /* First we process interests for specific metrics: uuid-metric|ST[tags...] */
+  for(i=0; i<2; i++) {
+    /* First time through we use the message's ID, second time we use the NULL UUID
+     * which is used to "match any" check.
+     */
+    if(mtev_hash_retrieve(&id_level, i ? (const char *)&uuid_zero : (const char *)&message->id.id, UUID_SIZE,
+        &vhash)) {
+      if(mtev_hash_retrieve((mtev_hash_table *) vhash, message->id.name,
+                            message->id.name_len_with_tags, (void **)&vinterests)) {
+        memcpy(interests, *vinterests, sizeof(interests)); /*  we access this exactly once, no need to ck load or tmp variable */
+        for(i=0; i<nthreads; i++) has_interests |= interests[i];
+      }
+      if(mtev_hash_retrieve((mtev_hash_table *) vhash, message->id.name,
+                            message->id.name_len, (void **)&vinterests)) {
+        memcpy(interests, *vinterests, sizeof(interests)); /*  we access this exactly once, no need to ck load or tmp variable */
+        for(i=0; i<nthreads; i++) has_interests |= interests[i];
```

Comment:
What's the visibility rule in C again for this?  Is this the same `i` than on line 426/435?

I believe the entire body of the for loop should be its own function anyway considering the importance of this code.